Greenrobot PR -- Ask questions to understand the code better DO NOT MERGE

habibmostafa opened the pull request:
@Koshkap @alpha4s @FarahAbdalla @ArshadurRahman @Ali100k 
Here is how you can ask questions to @Aman-Anas , @habibmostafa , @Abulila , @yousufbadar , @Github4Atif  and @Arshad-Anas.

- Go to https://github.com/raleighmasjid/marvels-ftc/pull/31
- Click on "Files changed"
- Read the code line by line and make sure you understand it.
- Click on plus sign next to the line you don't understand.
- Add a comment and tag someone you think knows the answer or don't tag.
- Click on "Add single comment".
<img width="652" alt="Screenshot 2023-01-22 at 9 48 52 PM" src="https://user-images.githubusercontent.com/1160465/213958427-e96bc80d-d9b6-4486-9b0a-628b076b7c6b.png">
- Resolve conversation if the answer makes sense to decrease clutter.
<img width="634" alt="Screenshot 2023-01-22 at 9 54 52 PM" src="https://user-images.githubusercontent.com/1160465/213958482-0c85f789-ebe4-4be7-bc00-dd9a8b956263.png">

habibmostafa committed:
add toggle claw

habibmostafa committed:
add toggle claw

habibmostafa committed:
add toggle claw

habibmostafa committed:
code for officiak scrimmage

habibmostafa committed:
code for official scrimmage

habibmostafa committed:
Update DriveTeleOp.java

Gamepad1 for drivetrain. Gamepad2 for slider and claw.

arshadanas committed:
Analog stick for lift, GreenBot extends mecanum drive

Switched slide controls from d-pad to analog stick
Separated mecanum drive and scoring functions, scoring functions (specific to green bot) are now in GreenBot class

arshadanas committed:
Added field-centric op mode

Field-relative teleOp mode

arshadanas committed:
Squared input for lift motors

Added comments to a bunch of stuff
Also removed gm0 field-centric drive code

arshadanas committed:
Added 'if' conditional for instantiating imu

This is to prevent confusion when coding autonomous

arshadanas committed:
Added Field Centric Reset Button

arshadanas committed:
Fieldcentric fixes

arshadanas committed:
fieldcentric

arshadanas committed:
IMU fixes

alpha4s committed:
12/24

alpha4s committed:
Merge branch 'greenrobot' of https://github.com/raleighmasjid/marvels-ftc into greenrobot

arshadanas committed:
moved init to top

arshadanas committed:
moved init to top

removed random scraps of unused code

arshadanas committed:
meepmeep testing

arshadanas committed:
First auton path

arshadanas committed:
Update DriveConstants.java

arshadanas committed:
Tuning done by the hardware people

Watch out, the hardware-people-made autonomous is running

arshadanas committed:
Added "greenBot" object to auton

arshadanas committed:
v1 gripper servo angles fix

arshadanas committed:
slide heights initial testing

arshadanas committed:
Fixed RR path issues at Elon

Scoring actions (PowerplayScorer class) now separate from holonomic drive actions (MarvelsMecanumDrive class); autonomous programs no longer create a MarvelsMecanumDrive object--was interfering with SampleMecanumDrive (RoadRunner)

arshadanas committed:
Added slide height controls

arshadanas committed:
Fixed slide height names

arshadanas committed:
april tags

April tag detection, telemetry, drop/liftClaw, PIDF setting

arshadanas committed:
Copied MeepMeep trajectory to auton opMode

alpha4s committed:
Yes

to the substation we go!

arshadanas committed:
Lift heights and lift encoder telemetry

arshadanas committed:
opmode naming fixes, 2-point auto update

alpha4s committed:
claw spin

arshadanas committed:
telemetry fixes and lift pid tuniung

arshadanas committed:
Ready for autonomous testing

Fully implemented passthrough movement (fast too) and preset slide heights (with decreased velocity on the way down)

arshadanas committed:
Programmed for first qualifier at Rolesville

21-point autonomous
Not async path following
Almost ready for autonomous--make async path following work somehow

arshadanas committed:
Changes at Rolesville

arshadanas committed:
Passthrough optimizations

Skip minor turn when lifted, quicker claw retraction, reduced time on moving claw to back

arshadanas committed:
Added async path following test

Also limit switch telemetry test

habibmostafa commented on MeepMeepTesting/src/main/java/com/example/meepmeeptesting/MeepMeepTesting.java:
```diff
@@ -0,0 +1,52 @@
+//                                .waitSeconds(0.1) // dropClaw.
+public class MeepMeepTesting {
+    public static void main(String[] args) {
+        MeepMeep meepMeep = new MeepMeep(650);
```
What is meepmeep?

Ali100k commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOpConfig.java:
```diff
@@ -3,10 +3,53 @@
 
 @Config
 public class TeleOpConfig {
-    public static double CLAW_LEFT_CLOSED = 0;
```
@Arshad-Anas What do the numbers next to the equal sign indicate?

habibmostafa replied:
Assalamu Alaikom Ali. No need to @ arshad specifically. He might have exams at this time and someone else can answer inshAllah.

Good question. So, this line is to assign a variable.
- CLAW_LEFT_CLOSED: The variable name 
- The value being assigned to it is 0.
- The variable is of type "double" (decimal basically)
- and it is public (meaning can be accessed from outside the class (TeleOpConfig))
- and it is static which means you can do something like TeleOpConfig. CLAW_LEFT_CLOSED anywhere in the code and it will work.

Resources to help you learn java in general and ask even better questions: https://books.trinket.io/thinkjava2/
Reach out on discord if you have questions on this: https://books.trinket.io/thinkjava2/

habibmostafa commented:
Resources:
What is roadrunner: https://learnroadrunner.com/#frequently-asked-questions

arshadanas committed:
First scoring autonomous

Also proper async path following, moved trajectory initialization to before init detection loop, renamed SampleMecanumDrive to AutonMecanumDrive

arshadanas committed:
Moved camera config to TeleOpConfig

arshadanas committed:
Working(-ish) autonomous scoring

almost consistent 1+1
Next fix: slow down on approach to medium junction to reduce slippage

arshadanas committed:
Constrained max velocity to reduce slippage

Fix stack grabbing accuracy issues from before

arshadanas committed:
Ready for final autonomous testing

Added back spline rotation at the start of autonomous
Constrained velocity when approaching junction and cone stack
Decreased claw closing timer length
Decreased down velocity on lift
Moved PIDF constant setting to before control loops
Replaced line path (for moving from preload scoring position to cone stack) with spline path
Simplified parking method
Added null tag checker to prevent NullPointerException crashes
Converted multiple positions and angles to variables for easy tuning
Rounded drive constants

arshadanas committed:
Manual lift control clipping and naming fixes

arshadanas committed:
Cleaned up init code and fixed possible major bug

Moved shared initialization statements to PowerplayScorer class
Renamed some variables for clarity

arshadanas committed:
Autonomous Improvements

habibmostafa committed:
Update AutonomousRight.java

arshadanas committed:
Working 1+3

arshadanas committed:
added testing class (working auton saved)

arshadanas committed:
Update AutonomousTesting.java

arshadanas committed:
Autonomous updates

Optimizations in active testing auton
Backup signal zone parking auton

arshadanas committed:
Camera rotation, precision mode, and voltage correction

arshadanas committed:
Increase manual lift control speed

Also removed unnecessary button reader

arshadanas committed:
Untested auton code (arshad r pls test)

arshadanas committed:
fixed auton + linear precision mode trigger

arshadanas committed:
New auton based on old with minor tweaks

arshadanas committed:
Transfer heading to teleop

arshadanas committed:
Overhauled passthrough control logic

arshadanas committed:
Naming, telemetry, and run statement fixes

arshadanas committed:
Straight line mode, longer waits for claw

Added 'straight mode' for driving in straight cardinal direction lines
Added 0.2 seconds for lifting cone slightly after grabbing
Set last junction scoring to 'dropClaw' and removed extra set lift position statement

arshadanas committed:
Working 1+3 v2

arshadanas committed:
Working 1+4 and adjusted passthru angles

arshadanas committed:
Adjusted pivot position

arshadanas committed:
Saved 1+4 as separate class

arshadanas committed:
Anti-drift adjust for zone 1 parking

arshadanas committed:
Adjusted constant sources to preserve versions

arshadanas committed:
Added last cycle for 1+5 attempt

+ telemetry updates

arshadanas committed:
BETTER ZONE 2 PARKING FINALLY

arshadanas committed:
Passthrough fixes

arshadanas committed:
Fixed 1+4

arshadanas committed:
Optimized for 1+5 i think we gonna do it

arshadanas committed:
1+4 revert fix mb

arshadanas committed:
added a 1+1 for aic practice

arshadanas committed:
Two toggle options for precision mode

arshadanas committed:
Update FieldRelativeTeleOp.java

arshadanas committed:
shared pose variable

arshadanas committed:
minor tweaks and fixes

- claw closed during init to hold cone
- removed field-centric toggle
- combined grab + lift and release + drop
- cleaned up precision mode code
- added override/failsafe statements to passthrough state machine

arshadanas committed:
constrained velocity at end of first path

arshadanas committed:
intentionally inconsistent 1+2 instead of 1+1

arshadanas committed:
Update AutonomousRight3Med.java

arshadanas committed:
Update FieldRelativeTeleOp.java

arshadanas committed:
Added limit switch functionality

arshadanas committed:
Update FieldRelativeTeleOp.java

arshadanas committed:
added tunable velocity to encoder reset

arshadanas committed:
Adjusted timers to prevent passthrough from hitting frame

arshadanas committed:
Comments in TeleOpConfig

arshadanas committed:
Naming fixes and removed useless timer

arshadanas committed:
Removed more useless timers

arshadanas committed:
Update PowerplayScorer.java

arshadanas committed:
Update TeleOpConfig.java

arshadanas committed:
at practice

arshadanas committed:
Update AutonomousTesting.java

arshadanas committed:
restored active testing

arshadanas committed:
Restored last saved working 1+4

arshadanas committed:
separating saved 1+4

arshadanas committed:
ADDED PASSTHROUGH SERVO SAVER

arshadanas committed:
servo saver v2

arshadanas committed:
servo saver update

arshadanas committed:
lift encoder reset enhancement

arshadanas committed:
servo saver update

arshadanas committed:
servo saver update

arshadanas committed:
Update PowerplayScorer.java

arshadanas committed:
Update TeleOpConfig.java

arshadanas committed:
Update PowerplayScorer.java

arshadanas committed:
Added lift PIDF override

arshadanas committed:
updated limit switch reset

arshadanas committed:
Minor adjustments

arshadanas committed:
Servo saver and lift override updates

arshadanas committed:
Naming updates
